editors/code/src/debug.ts

```diff
@@ -359,3 +370,56 @@ function quote(xs: string[]) {
         })
         .join(" ");
 }
+
+const execShell = (cmd: string) => {
+    const root = vscode.workspace.workspaceFolders
+        ? vscode.workspace.workspaceFolders[0]?.uri.path
+        : undefined;
+
+        exec(cmd, { cwd: root }, (error, stdout, stderr) => {
+        if (stdout) {
+            log.info(stdout);
+        }
+        if (stderr) {
+            log.error(`Error: ${stderr}`);
+        }
+        if (error) {
+            log.error("Failed to run command: " + error.message);
+        }
+    });
+};
+
+export function initializeDebugSessionTrackingAndRebuild() {
+    vscode.debug.onDidStartDebugSession((session: vscode.DebugSession) => {
+        if(!activeDebugSessionIds.includes(session.id)) {
+            activeDebugSessionIds.push(session.id);
+        }
+        
+        if(!debugSessionIdBuffer.includes(session.id)) {
+            debugSessionIdBuffer.push(session.id);
+        }
+    });
+
+    vscode.debug.onDidTerminateDebugSession(async (session: vscode.DebugSession) => {
+        cleanupDebugSessionIdBuffer();
+        removeActiveSession(session);
+    
+        // The id of the session will be the same when pressing restart the restart button
+        if (debugSessionIdBuffer.find((s) => s === session.id)) {
+            execShell("cargo test --no-run");
```

Comment:
why are we executing cargo when the session ends?

Reply:
This takes care of the building process before the new test begins. Unfortunately, doing that in the `onDidStartDebugSession` hook is too late.